crypto/mem.c

```diff
@@ -86,6 +94,18 @@ void CRYPTO_get_mem_functions(
 }
 
 #ifndef OPENSSL_NO_CRYPTO_MDEBUG
+void CRYPTO_get_alloc_counts(int *mcount, int *rcount, int *fcount)
+{
+    CRYPTO_THREAD_write_lock(malloc_lock);
```

Comment:
You do realize that CRYPTO_atomic_add(&x, 1, &dummy, malloc_lock) doesn't use malloc_lock on contemporary platforms. I mean the referred lock doesn't prevent rest of the code from modifying values. On contemporary platforms. In other words if accurate counters are all-important, then one shouldn't use CRYPTO_atomic_add.

Reply:
It's the best we have, isn't it?  Is there any other portable choice for atomic increment?

Reply:
> It's the best we have, isn't it?

What do you mean? What I'm trying to say is that maybe it's not right application for CRYPTO_atomic_add, in which case it doesn't matter if it's best or not. And if you can't use it, then you've obliged to do it the "usual way", i.e. lock, counter++, unlock. If we insist of atomic, then there would have to be atomic_compare_exchange, that would *zero* corresponding counter. So that there could be only one CRYPTO_get_alloc_counts caller and would be its responsibility to accumulate values if it calls CRYPTO_get_alloc_counts multiple times. Alternatively you can always say that in MEM_DEBUG scenario there will one call at the end, after everything is done, so it doesn't matter if it's locked or not...

Reply:
I am not worried about multiple threads doing simultaneous allocation operations and the counts being off by a few.  If someone has a better definition of INCREMENT for me to put in this PR, great.

Reply:
I find it jarring as a reader when the formal thread-safety rules are not adhered to (at least without a comment attempting to explain the rule violation).  Not mixing direct accesses to a variable and CRYPTO_atomic_add is one such rule.

Reply:
so you like the code as-is, @kaduk?

Reply:
> so you like the code as-is

I can't quite tell what you're actually asking about, so I'll be a little flippant and note that there are some parts of the OpenSSL master source tree that I like.  The current state of this pull request falls under the "jarring to read" category, though.

Reply:
What do you think this PR is doing wrong?  The values are set and fetched under the malloc_lock, even if some implementations don't use that lock.

Reply:
> The values are set and fetched under the malloc lock

But they are not set under the malloc_lock!  INCREMENT(), aka.  ```CRYPTO_atomic_add(&x, 1, &dummy, malloc_lock)``` expands to ```dummy = __atomic_add_fetch(&x, 1, __ATOMIC_ACQ_REL)``` on "most modern" systems, without ever acquiring malloc_lock.  So, holding a write lock on malloc_lock does not block any concurrent INCREMENT calls.